engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/PrivateAccessTest.java

```diff
@@ -0,0 +1,159 @@
+            private Cons data
+        main = My_Type.Cons 42
+        """;
+    var projDir = createProject("My_Project", mainSrc);
+    var mainSrcPath = projDir.resolve("src").resolve("Main.enso");
+    try (var ctx =
+        defaultContextBuilder()
+            .option(RuntimeOptions.PROJECT_ROOT, projDir.toAbsolutePath().toString())
+            .build()) {
+      var polyCtx = new PolyglotContext(ctx);
+      var mainMod = polyCtx.evalModule(mainSrcPath.toFile());
+      var assocType = mainMod.getAssociatedType();
+      var mainMethod = mainMod.getMethod(assocType, "main").get();
+      var res = mainMethod.execute();
+      assertThat(res.hasMember("data"), is(false));
+      try {
+        res.getMember("data");
+        fail("Expected exception in getMember('data')");
+      } catch (UnsupportedOperationException e) {
+        // nop - expected
+      }
+    }
+  }
+
+  @Test
+  public void privateConstructorIsNotExposedToPolyglot() throws IOException {
+    var mainSrc = """
+        type My_Type
+            private Cons data
+        """;
+    var projDir = createProject("My_Project", mainSrc);
+    var mainSrcPath = projDir.resolve("src").resolve("Main.enso");
+    try (var ctx =
+        defaultContextBuilder()
+            .option(RuntimeOptions.PROJECT_ROOT, projDir.toAbsolutePath().toString())
+            .build()) {
+      var polyCtx = new PolyglotContext(ctx);
+      var mainMod = polyCtx.evalModule(mainSrcPath.toFile());
+      var myType = mainMod.getType("My_Type");
+      assertThat(myType.hasMember("Cons"), is(false));
+    }
+  }
+
+  /** Tests that pattern matching on private constructors fails in compilation. */
+  @Test
+  public void cannotPatternOnMatchPrivateConstructor() throws IOException {
+    var libSrc =
+        """
+        type My_Type
+            private Cons data
+            create x = My_Type.Cons x
+        """;
+    createProject("Lib", libSrc);
+    var projSrc =
+        """
+        from local.Lib import My_Type
+        main =
+            obj = My_Type.create 42
+            case obj of
+                My_Type.Cons x -> x
+        """;
+    var projDir = createProject("Proj", projSrc);
+    var out = new ByteArrayOutputStream();
+    try (var ctx =
+        defaultContextBuilder()
+            .option(RuntimeOptions.PROJECT_ROOT, projDir.toAbsolutePath().toString())
+            .option(RuntimeOptions.STRICT_ERRORS, "true")
+            .option(RuntimeOptions.DISABLE_IR_CACHES, "true")
+            .out(out)
+            .err(out)
+            .build()) {
+      var polyCtx = new PolyglotContext(ctx);
+      try {
+        polyCtx.getTopScope().compile(true);
+        fail("Expected compiler error");
+      } catch (PolyglotException e) {
+        assertThat(
+            out.toString(),
+            allOf(
+                containsString("error:"),
+                containsString("Project-private constructor"),
+                containsString("cannot be used from")));
```

Comment:
I added a comment outside of PR: https://github.com/enso-org/enso/commit/492f3093399a91c34201f5db9d7316a4ce7e156c#r141262725

I think it would be beneficial to make this check a bit more specific, at least changing it to a single `containsString` that 'shows' the more of the error message, eg.:
```suggestion
                containsString("Project-private constructor My_Type.Cons in project ?? cannot be used from ?WHERE?"),
```

Reply:
I don't like tests that have to be updated with every code change. Checking exact error message looks like one of such tests. I prefer checking only for the important parts of the message as the
```java
                containsString("error:"),
                containsString("Project-private constructor"),
                containsString("cannot be used from")));
```
seem to do.

Reply:
Okay.
Let's then at least add the constructor to the tests - I'd like to know if it's just the constructor name or constructor name together with type name (and ideally I think it should be the latter so that the user can more easily identify it).

Reply:
You can see the formatting of the error message in [Standard.Base.Errors.Common.Private_Access](https://github.com/enso-org/enso/blob/85dca02f0d8f14e03c984b6425cd624b702d6664/distribution/lib/Standard/Base/0.0.0-dev/src/Errors/Common.enso#L311) - it contains caller project name, callee project name and the name of the method. I think that is sufficient information.

In practice it looks like this:
```
Execution finished with an error: Private access error: The project-private method 'local.Lib.Main.My_Type.Cons' in project 'Lib' cannot be accessed from project 'Proj'.
        at <enso> cons(src/Main.enso:6:14-27)
        at <enso> Main.main(src/Main.enso:7:5-11)
```

Reply:
I see, thanks for the example.

One remark though: it would be good to include the _namespace_ in the project name here as well.

Imagine you have projects `Standard.Table` and `Radeusgd.Table`. Then the error will look like:
```
Execution finished with an error: Private access error: The project-private method 'Standard.Table.Main.My_Type.Cons' in project 'Table' cannot be accessed from project 'Table'.
```

While each projects can be inferred from context (the caller is from the source code location and callee from the fully qualified method name), it is a bit confusing to see the fragment `(method) in project 'Table' cannot be accessed from project 'Table'`. So IMO it would be best to turn this into:
```
Execution finished with an error: Private access error: The project-private method 'Standard.Table.Main.My_Type.Cons' in project 'Standard.Table' cannot be accessed from project 'Radeusgd.Table'.
```